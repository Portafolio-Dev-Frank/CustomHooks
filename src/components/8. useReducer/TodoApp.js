import React, { useEffect, useReducer } from 'react'
import './styles.css'
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => {
    // return [{
    //     id: new Date().getTime(),
    //     desc: 'aprender React',
    //     done: false
    // }]
    return JSON.parse(sessionStorage.getItem('todos')) || [];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);


    useEffect(() => {
        sessionStorage.setItem('todos', JSON.stringify(todos))
        return () => {
            console.log('removido el effect');
        }
    }, [todos])


    const handleAdd = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    const handleDelete = (todoId) => {
        dispatch({
            type: 'remove',
            payload: todoId
        })
    }

    const handleToggle = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId
        })
    }



    return (
        <div>
            <h1>TodoApp : ({todos.length})</h1>
            <hr></hr>
            <div className="row">
                <div className="col-7">
                    <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete}></TodoList>
                </div>
                <div className="col-5">
                    <TodoAdd handleAdd={handleAdd}></TodoAdd>
                </div>

            </div>
        </div>
    )
}
