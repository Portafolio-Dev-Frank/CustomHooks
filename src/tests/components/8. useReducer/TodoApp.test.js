import { act } from "@testing-library/react"
import { mount, shallow } from "enzyme"
import { TodoApp } from "../../../components/8. useReducer/TodoApp"
import { demosTodos } from "../../fixtures/demoTodos"

describe('pruebas en componente <TodoApp></TodoApp>', () => {


    const wrapper = shallow(<TodoApp></TodoApp>)

    Storage.prototype.setItem = jest.fn(() => {});

    test('debe de renderizarse el componente correctamente ', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de agregar un TODO', () => {
        // mount mas complejo que shallow
        const wrapper = shallow(<TodoApp></TodoApp>);

        act(() => {
            wrapper.find('TodoAdd').prop('handleAdd')(demosTodos[0]);
            wrapper.find('TodoAdd').prop('handleAdd')(demosTodos[1]);
        })
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp : (2)');
        // expect(localStorage.setItem).toHaveBeenCalledTimes(3);
    })

})
