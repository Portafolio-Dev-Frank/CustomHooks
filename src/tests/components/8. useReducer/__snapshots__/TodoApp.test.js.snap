// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en componente <TodoApp></TodoApp> debe de renderizarse el componente correctamente  1`] = `
<div>
  <h1>
    TodoApp : (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
